tests/activation/test_prompts.py

```diff
@@ -0,0 +1,349 @@
+"""test that prompt behavior is correct in supported shells"""
+from __future__ import absolute_import, unicode_literals
+
+import os
+import subprocess
+import sys
+from collections import namedtuple
+from textwrap import dedent
+
+import pytest
+
+import virtualenv
+from virtualenv import IS_DARWIN, IS_WIN
+
+try:
+    from pathlib import Path
+except ImportError:
+    from pathlib2 import Path
+
+VIRTUAL_ENV_DISABLE_PROMPT = "VIRTUAL_ENV_DISABLE_PROMPT"
+
+# This must match the DEST_DIR provided in the ../conftest.py:clean_python fixture
+ENV_DEFAULT = "env"
+
+# This can be anything
+ENV_CUSTOM = "env_custom"
+
+# Standard prefix, surround the env name in parentheses and separate by a space
+PREFIX_DEFAULT = "({}) ".format(ENV_DEFAULT)
+
+# Arbitrary prefix for the environment that's provided a 'prompt' arg
+PREFIX_CUSTOM = "---ENV---"
+
+# Temp script filename template: {shell}.script.(normal|suppress).(default|custom)[extension]
+SCRIPT_TEMPLATE = "{}.script.{}.{}{}"
+
+# Temp output filename template: {shell}.out.(normal|suppress).(default|custom)
+OUTPUT_TEMPLATE = "{}.out.{}.{}"
+
+SHELL_LIST = ["bash", "fish", "csh", "xonsh", "cmd", "powershell"]
+
+
+# Py2 doesn't like unicode in the environment
+def env_compat(string):
+    return string.encode("utf-8") if sys.version_info.major < 3 else string
+
+
+@pytest.fixture(scope="module")
+def posh_execute_enabled(tmp_path_factory):
+    if not IS_WIN:
+        return False
+
+    test_ps1 = tmp_path_factory.mktemp("posh_test") / "test.ps1"
+    with open(str(test_ps1), "w") as f:
+        f.write("echo foo\n")
+
+    return 0 == subprocess.call("powershell -File {}".format(str(test_ps1)), shell=True)
+
+
+@pytest.fixture(scope="module")
+def platform_check_skip(posh_execute_enabled, pytestconfig):
+    """Check whether to skip based on platform & shell.
+
+    Returns a string if test should be skipped, or None if test should proceed.
+
+    """
+    platform_incompat = "No sane provision for {} on {} yet"
+
+    def check(platform, shell):
+
+        if shell == "bash":
```

Comment:
feels to me like a lot of code repetition here? Can't we make it shorter by some inheritance/class attributes as we did in the other activation test suite?

Reply:
Hmm... so, say -- folding `platform_check_skip` together with `shell_info` and refactoring into classes?

Oh, and, probably [this per-shell customization](https://github.com/pypa/virtualenv/blob/a46be0f90ccfd65aeefc8cb500851b4d6960d106/tests/activation/test_prompts.py#L338-L349) of `test_activated_prompt` could go in there too?

Reply:
Yeah sounds good to me  👌

Reply:
Given the above two changes I'm comfortable with merging this in, thanks for the contribution! Let me know once you finish them.

Reply:
Ok, @gaborbernat -- first attempt at the refactor into the shell-info class hierarchy is implemented in 677bbc1. Thoughts please!